docs/proposals/petset.md

```diff
@@ -0,0 +1,533 @@
+  than running a clustered system off Kube. Authors should be given tools to guard against common cluster
+  failure modes (split brain, phantom member) to prevent introducing more failure modes. Sophisticated
+  distributed systems designers can implement more sophisticated solutions than PetSet if necessary -
+  new users should not become vulnerable to additional failure modes through an overly flexible design.
+* **Limited scaling** - While flexible scaling is important for some clusters, other examples of clusters
+  do not change scale without significant external intervention. Human intervention may be required after
+  scaling. Changing scale during cluster operation can lead to split brain in quorum systems. It should be
+  possible to scale easily, but the details of making that safe belong to the pods and image authors.
+* **No generic cluster lifecycle** - Rather than design a general purpose lifecycle for clustered software,
+  focus on ensuring the information necessary for the software to function is available. For example,
+  rather than providing a "post-creation" hook invoked when the cluster is complete, provide the necessary
+  information to the "first" (or last) pod to determine the identity of the remaining cluster members and
+  allow it to manage its own initialization.
+* **External access direct to cluster members is out of scope** - exposing pods to consumers that cannot
+  access the pod network is out of scope, because we do not currently support headless services being
+  exposed via NodePort. A workaround is to allow external clients to access the pod network, or to
+  create one NodePort service per member. A future design should cover headless external service access.
+
+
+## Proposed Design
+
+Add a new resource to Kubernetes to represent a set of pods that are individually distinct but each
+individual can safely be replaced-- the name **PetSet** (working name) is chosen to convey that the
+individual members of the set are themselves "pets" and thus each one is important. A relevant analogy
+is that a PetSet is composed of pets, but the pets are like goldfish. If you have a blue, red, and
+yellow goldfish, and the red goldfish dies, you replace it with another red goldfish and no one would
+notice. If you suddenly have three red goldfish, someone will notice.
+
+The PetSet is responsible for creating and maintaining a set of **identities** and ensuring that there is
+one pod and zero or more **supporting resources** for each identity. There should never be more than one pod
+or unique supporting resource per identity at any one time. A new pod can be created for an identity only
+if a previous pod has been fully terminated (reached its graceful termination limit or cleanly exited).
+
+```
+type PetSet struct {
+    ObjectMeta
+
+    Spec PetSetSpec
+    ...
+}
+
+type PetSetSpec struct {
+    // Number of instances in the set
+    Replicas *int
```

Comment:
PeerSet?  Peers are equal in importance but have their own identity.
(Forget any network connotations and think about people.).
On Feb 16, 2016 7:47 PM, "Clayton Coleman" <notifications@github.com> wrote:

> In docs/proposals/petset.md
> <https://github.com/kubernetes/kubernetes/pull/18016#discussion_r53118024>
> :
>
> > +The PetSet is responsible for creating and maintaining a set of **identities** and ensuring that there is
> > +one pod and zero or more **supporting resources** for each identity. There should never be more than one pod
> > +or unique supporting resource per identity at any one time. A new pod can be created for an identity only
> > +if a previous pod has been fully terminated (reached its graceful termination limit or cleanly exited).
> > +
> > +```
> > +type PetSet struct {
> > +    ObjectMeta
> > +
> > +    Spec PetSetSpec
> > +    ...
> > +}
> > +
> > +type PetSetSpec struct {
> > +    // Number of instances in the set
> > +    Replicas *int
>
> I'm still not comfortable with the idea of having a "Set Template" - true,
> it's a more generic tool that can be used to accomplish many things, but on
> the other hand it's a more generic tool that can be used to accomplish many
> things that client programs have no way to reason about their intent. That
> I think is the strongest value in having a formal type here - providing a
> set of strong guarantees that represent the needs of clustered software
> discussed in this issue - a focus on solving a specific problem, rather
> than a generic one. Having the controller be generic to the set members
> means that we have no strong control over the semantics of the pods under
> the set, and by using replicasets at each instance we also lose the "at
> most one" and stable name guarantees that make the pods easier to reason
> about. It formalizes our existing workaround... but our existing workaround
> doesn't work well for cluster members that should be "at most 1" during
> normal operation (until the pod is completely gone). It does mean we get to
> reuse our existing RC function - but that is another layer of object to
> reason about for client programs. Today, looking at our API, I could
> somewhat hazard what I should use Jobs, ReplicaSets, DaemonSets, and
> Deployments for. Our other objects exist to solve use cases that in some
> ways relate to running software, or connecting / enabling running software.
> This would be the first "meta" resource that exists to clone anything /
> everything. We had discussed having each controller manage its own updates
> or have a related object to do it. A ParameterizedSet seems like it would
> have to pick a set of generic update patterns, or be truly generic, but not
> be able to strongly reason about the resources under its control. My
> preference would be to only introduce ParameterizedSet if we think we
> needed strong stable identity for lots of different types of resources or
> we felt that we need a generic tool to clone resources. I don't think we've
> felt that yet - IndexedJob seems easier to reason about as a user than
> implementing IndexedJob with ParameterizedSet. Likewise a PetSet seems
> easier to reason about as a user than a ParameterizedSet of ReplicaSets. As
> originally envisioned, the goal here was to introduce the lower level
> resource that is the equivalent of ReplicaSet but for stable identity
> objects. Perhaps the really compelling argument would come down to whether
> I would use a ParameterizedSet for DaemonSets, or for IndexedJobs, or for
> PetSets. If so, maybe ParameterizedSet is really just a distinct resource
> that should exist at a higher level, and possibly related to the concept
> discussed on the Ubernetes working group for generic replication across
> clusters (where each set member might be on a different cluster).
> … <#149003185_>
> On Tue, Feb 16, 2016 at 8:58 PM, Clayton Coleman <ccoleman@redhat.com>
> wrote: Some other ones suggested recently by folks: StableSet OrderedSet
> IdentitySet MemberSet I do like PersistentSet since the primary goal is to
> assign unique volumes to each identity.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/18016/files#r53118024>.
>